src/topology/local_homeomorph.lean

```diff
@@ -0,0 +1,479 @@
+  open_target        := e.open_source,
+  continuous_to_fun  := e.continuous_inv_fun,
+  continuous_inv_fun := e.continuous_to_fun,
+  ..e.to_local_equiv.symm }
+
+@[simp] lemma symm_to_local_equiv : e.symm.to_local_equiv = e.to_local_equiv.symm := rfl
+@[simp] lemma symm_to_fun : e.symm.to_fun = e.inv_fun := rfl
+@[simp] lemma symm_inv_fun : e.symm.inv_fun = e.to_fun := rfl
+@[simp] lemma symm_source : e.symm.source = e.target := rfl
+@[simp] lemma symm_target : e.symm.target = e.source := rfl
+@[simp] lemma symm_symm : e.symm.symm = e := eq_of_local_equiv_eq $ by simp
+
+/-- Preimage of interior or interior of preimage coincide for local homeomorphisms, when restricted
+to the source. -/
+lemma preimage_interior (s : set β) :
+  e.source ∩ e.to_fun ⁻¹' (interior s) = e.source ∩ interior (e.to_fun ⁻¹' s) :=
+begin
+  apply subset.antisymm,
+  { exact e.continuous_to_fun.preimage_interior_subset_interior_preimage e.open_source },
+  { calc e.source ∩ interior (e.to_fun ⁻¹' s)
+    = (e.source ∩ e.to_fun ⁻¹' e.target) ∩ interior (e.to_fun ⁻¹' s) : begin
+        congr,
+        apply (inter_eq_self_of_subset_left _).symm,
+        apply e.to_local_equiv.source_subset_preimage_target,
+      end
+    ... = (e.source ∩ interior (e.to_fun ⁻¹' s)) ∩ (e.to_fun ⁻¹' e.target) :
+      by simp [inter_comm, inter_assoc]
+    ... = (e.source ∩ e.to_fun ⁻¹' (e.inv_fun ⁻¹' (interior (e.to_fun ⁻¹' s)))) ∩ (e.to_fun ⁻¹' e.target) :
+      by rw e.to_local_equiv.source_inter_preimage_inv_preimage
+    ... = e.source ∩ e.to_fun ⁻¹' (e.target ∩ e.inv_fun ⁻¹' (interior (e.to_fun ⁻¹' s))) :
+       by rw [inter_comm e.target, preimage_inter, inter_assoc]
+    ... ⊆ e.source ∩ e.to_fun ⁻¹' (e.target ∩ interior (e.inv_fun ⁻¹' (e.to_fun ⁻¹' s))) : begin
+        apply inter_subset_inter (subset.refl _) (preimage_mono _),
+        exact e.continuous_inv_fun.preimage_interior_subset_interior_preimage e.open_target
+      end
+    ... = e.source ∩ e.to_fun ⁻¹' (interior (e.target ∩ e.inv_fun ⁻¹' (e.to_fun ⁻¹' s))) :
+      by rw [interior_inter, interior_eq_of_open e.open_target]
+    ... = e.source ∩ e.to_fun ⁻¹' (interior (e.target ∩ s)) :
+      by rw e.to_local_equiv.target_inter_inv_preimage_preimage
+    ... = e.source ∩ e.to_fun ⁻¹' e.target ∩ e.to_fun ⁻¹' (interior s) :
+      by rw [interior_inter, preimage_inter, interior_eq_of_open e.open_target, inter_assoc]
+    ... = e.source ∩ e.to_fun ⁻¹' (interior s) : begin
+        congr,
+        apply inter_eq_self_of_subset_left,
+        apply e.to_local_equiv.source_subset_preimage_target
+      end }
+end
+
+/-- Restricting a local homeomorphism `e` to `e.source ∩ interior s`. We use the interior to make
+sure that the restriction is well defined whatever the set s, since local homeomorphisms are by
+definition defined on open sets. In applications where `s` is open, this coincides with the
+restriction of local equivalences -/
+protected def restr (s : set α) : local_homeomorph α β :=
```

Comment:
Can't you define this as `e.restr_open (interior s) _` to save some duplicate work?

Reply:
Done